ADDED: library(crypto): New interface predicates for computing hashes

triska opened the pull request:
The new predicates `crypto_term_hash/3`, `crypto_file_hash/3` etc. finally make the digest&nbsp;algorithms that are currently available in library(md5) and library(sha) available under a unified interface, by providing bindings to OpenSSL functionality.

Further OpenSSL digest algorithms can be made available easily with this interface.

Please review especially the Prolog **interface** predicates provided by this patch.

I think this interface is simpler than that of `library(sha)`, for example<sup>1</sup>.

These interface predicates subsume existing libraries that ship with SWI and make it easy to provide new digest algorithms in the future. For example, one of the first new additions will be the `BLAKE` hashing algorithms that are available since OpenSSL&nbsp;1.1.0c. I will include them in `library(crypto)` as soon as https://github.com/openssl/openssl/issues/2169 is resolved, i.e., their **names** are finalized in&nbsp;OpenSSL.

I would be grateful for any review. If you like it, I think this will be ready for merging by the end of the week. Until then, I may update the pull request with minor improvements and corrections of all issues I find in the coming&nbsp;days.

----
<sup>1</sup> This is also evidenced by the fact that `library(sha)` itself was unable to use its own interface correctly: For example, the *empty* file *fails* for `file_sha1(EmptyFile, Hash)`, whereas you correctly get with `library(crypto)`:

<pre>
?- crypto_file_hash(empty_file, Hash, <b>[algorithm(sha1)]</b>).
<b>Hash = 'da39a3ee5e6b4b0d3255bfef95601890afd80709'.</b>
</pre>

JanWielemaker commented on crypto.pl:
```diff
@@ -33,7 +33,12 @@
 
 :- module(crypto,
-          [ evp_decrypt/6,              % +CipherText, +Algorithm, +Key, +IV, -PlainText, +Options
+          [ crypto_term_hash/3,         % +Data, -Hash, +Options
```
The name is potentially confusing with the widely accepted term_hash/2,4 family, while _data_ can only be text.  crypto_string_hash/3?

triska replied:
I am now using `crypto_data_hash/3`, since the arguments were also denoted as `Data` in several other predicates, and it can be a string, atom or list of characters.

wouterbeek replied:
Is it customary to refer to atom|string|chars|codes as 'data'?  A number is also data.  A stream and a file also contain data.

BTW the documentation of `crypto_data_hash/3` talks about a "code-list".  This is not the same as "a list of characters" or are both supported?

triska replied:
Code-lists are the least elegant way to represent strings, and they are *also* supported. Lists of characters are a good practical choice and *portable* with `double_quotes` set to `chars`. I note *portable* explicitly because *strings* are&nbsp;*not*.

"Data" is used in the documentation of these arguments already in `library(md5)` and `library(sha)`. "Data" never refers to a stream or file handle in the existing documentation.

triska replied:
Personally, I would still prefer `crypto_term_hash/3`, with the prospect to extend this in the future to more terms. I originally chose this term exactly because `term_hash/2` is already known. However, `crypto_data_hash/3` is also sufficiently general to allow&nbsp;this, and matches the variable names that are used in the documentation of the existing hash predicates to refer to the&nbsp;data.

JanWielemaker commented on crypto4pl.c:
```diff
@@ -66,6 +68,255 @@ typedef enum
+typedef struct context
+{ int             magic;
+  atom_t          atom;
+  unsigned int    encoding;
```
encoding is of type `IOENC`

triska replied:
I used `md54pl.c` as an implementation example. Changed in the latest update of this pull request.

JanWielemaker commented on crypto4pl.c:
```diff
@@ -66,6 +68,255 @@ typedef enum
+
+typedef struct context
+{ int             magic;
+  atom_t          atom;
```
?? _atom_ is a hard to interpret member name.  Please change or comment.

triska replied:
I used `ssl4pl.c` as an implementation example. To me, this member is one of the few that made immediate sense when I saw the library: It is the atom registered for this context. I cannot think of a better name to denote an atom. However, if you want to change it, please also change it in the other libraries where this technique is used.

JanWielemaker replied:
Ok.  This was Matt's work.  I normally put a little comment with struct members and typically use `symbol` for this one.  Don't bother.

JanWielemaker commented on crypto4pl.c:
```diff
@@ -66,6 +68,255 @@ typedef enum
+  result->encoding = REP_UTF8;
+  result->algorithm = EVP_sha256();
+
+  while(PL_get_list(opts, opt, opts))
```
If you use PL_get_list_ex and PL_get_nil_ex and simply return FALSE if they fail.

triska replied:
Thank you, I have adjusted this.

wouterbeek commented:
Thanks for working on this!  A few questions:

  - What is your definition of a context?  It is used in several predicates but I do not know what it means.

  - While it is good to have wrapper predicates for specific by common use cases like calculating the hash of an atom or (the contents of) a file, it would be nice to have 2 public predicates that support the generic use case of (incrementally) calculating the hash of a stream: (1) open a new stream for calculating the hash of an existing stream (like `zopen/3` and `open_hash_stream/3`), (2) retrieve the (incrementally) calculated hash for such an opened stream (like `stream_hash/2`).

  - What is the relation between this library, `hash_stream`, `md5` and `sha`?

triska commented:
@wouterbeek: Thank you for looking into this.

To answer your questions in order (please note that I am referring to the now *updated* pull request, where a few predicate names have slightly changed):

- A **context** for hashing is created with `crypto_context_new/2`. You can then continuously update it with `crypto_data_context/4`, and at any time fetch the current hash with `crypto_context_hash/2`.
- I have added `crypto_open_hash_stream/3` and `crypto_stream_hash/2`, please have a look.
- `library(md5)` and `library(sha)` both fall short in several ways. For example, `library(md5)` lacks stream-based hashing. Conversely, `library(sha)` lacks MD5 hashing. Neither of these libraries is a good fit for hosting additional algorithms that will become available or are already available, such as `BLAKE` (available with OpenSSL&nbsp;1.1.0), since their library names are not general enough. Also, their interfaces are not uniform, so that users always have to draw a case distinction between MD5 and SHA when using these libraries. Further, the default algorithm of `library(sha)` is becoming [**obsolete**](https://blog.qualys.com/ssllabs/2014/09/09/sha1-deprecation-what-you-need-to-know) and advised against.

The existing libraries already point to OpenSSL in their documentation:

- `md5`: "This  is a rather short-term solution waiting for a more general interface to the libcrypto functions of OpenSSL."
- `hash_stream`: "In addition to this hash library, SWI-Prolog provides `library(md5)`, `library(sha)` and hash functions through `library(ssl).`.


Therefore, the new hashing predicates of `library(crypto)` are designed and implemented to **fully&nbsp;subsume** and extend *each* of the hash libraries (`md5`, `sha`, `hash_stream`) that currently ship with SWI-Prolog, and make all the existing functionality available with a more general and more convenient interface that lets us fully harness *all* digest algorithms that are provided by&nbsp;OpenSSL.

Please try the updated pull request and let me know if everything works for you. Thank you!

wouterbeek commented:
@triska Thank you for the additions/clarifications.  I like your vision of unifying+extending many specific modules with one more generic module.  Feature-wise `crypto` seems to fully subsume `sha` and `hash_stream`, but can you also confirm that this is the case in other respects, e.g., performance-wise?  Is calculating a hash incrementally with `crypto` as fast or faster than doing the same with `hash_stream`?

triska commented:
@wouterbeek: Performance is roughly the same  in the cases I tested, with at most a few percent difference between `library(crypto)` and `library(hash_stream)` for a 1GB&nbsp;file. I would be grateful if you could also run a few&nbsp;tests.

JanWielemaker commented:
> Therefore, the new hashing predicates of library(crypto) are designed and implemented to fully subsume and extend each of the hash libraries (md5, sha, hash_stream) that currently ship with SWI-Prolog, and make all the existing functionality available with a more general and more convenient interface that lets us fully harness all digest algorithms that are provided by OpenSSL.

While I do like a generic interface, the size and dependencies of OpenSSL justify (IMO) the existence of small and stand alone alternatives.    We should highlight this relation in the docs.

JanWielemaker commented:
@triska Is this branch ready for final checking and merging?

triska commented:
Yes, it is ready from my perspective. Please update the related documentation in the other hash libraries as you see fit. In my view, OpenSSL will be increasingly used in the future in SWI web applications anyways, and so in fact loading the *other* libraries could increase, even if only slightly, the memory footprint over sticking to `library(crypto)` while also limiting the available algorithms. I hope you also take this into account when recommending one over the other.
